File: forge-core/internal/kerneloperationalcontract/records.go
package kerneloperationalcontract

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"reflect"
)

func canonicalTyped(value any, maximum int) ([]byte, error) {
	node, err := typedNode(value, maximum)
	if err != nil {
		return nil, err
	}
	return canonicalJSON(node, maximum)
}

func typedDigest(value any, domain string, maximum int) (string, error) {
	canonical, err := canonicalTyped(value, maximum)
	if err != nil {
		return "", err
	}
	hasher := sha256.New()
	_, _ = hasher.Write([]byte(domain))
	_, _ = hasher.Write(canonical)
	return hex.EncodeToString(hasher.Sum(nil)), nil
}

func cloneValue[T any](value *T) (*T, error) {
	// encoding/json replaces invalid UTF-8 in Go strings with U+FFFD. Validate
	// the caller-owned typed tree before marshaling so cloning can never turn a
	// rejected wire value into a different, valid value.
	if err := validateTypedValue(reflect.ValueOf(value), 1); err != nil {
		return nil, err
	}
	data, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	var result T
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

func artifactReceiptDigest(value *ArtifactReceipt) (string, error) {
	blank, err := cloneValue(value)
	if err != nil {
		return "", err
	}
	blank.ArtifactReceiptID, blank.ArtifactReceiptSHA256 = "", ""
	if err := validateArtifactReceiptFields(blank, true); err != nil {
		return "", err
	}
	return typedDigest(blank, artifactReceiptDomain, maxArtifactReceiptBytes)
}

func validateArtifactReceipt(value *ArtifactReceipt) error {
	if err := validateArtifactReceiptFields(value, false); err != nil {
		return err
	}
	digest, err := artifactReceiptDigest(value)
	if err != nil {
		return err
	}
	if value.ArtifactReceiptSHA256 != digest {
		return fmt.Errorf("artifact_receipt_sha256 does not match canonical preimage")
	}
	_, err = canonicalTyped(value, maxArtifactReceiptBytes)
	return err
}

// SealArtifactReceipt seals one exact blank-identity ArtifactReceipt copy.
func SealArtifactReceipt(value *ArtifactReceipt) (*ArtifactReceipt, error) {
	if value == nil || value.ArtifactReceiptID != "" || value.ArtifactReceiptSHA256 != "" {
		return nil, fmt.Errorf("sealing ArtifactReceipt requires blank identity fields")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := artifactReceiptDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.ArtifactReceiptID = artifactReceiptPrefix + digest
	sealed.ArtifactReceiptSHA256 = digest
	return sealed, validateArtifactReceipt(sealed)
}

// DecodeArtifactReceipt decodes one exact canonical sealed ArtifactReceipt.
func DecodeArtifactReceipt(data []byte) (*ArtifactReceipt, error) {
	var value ArtifactReceipt
	if err := decodeTypedExact(data, maxArtifactReceiptBytes, &value); err != nil {
		return nil, err
	}
	return &value, validateArtifactReceipt(&value)
}

func invocationDigest(value *CapabilityInvocation) (string, error) {
	blank, err := cloneValue(value)
	if err != nil {
		return "", err
	}
	blank.InvocationID, blank.InvocationSHA256 = "", ""
	if err := validateInvocationFields(blank, true); err != nil {
		return "", err
	}
	return typedDigest(blank, invocationDomain, maxInvocationBytes)
}

func validateInvocation(value *CapabilityInvocation) error {
	if err := validateInvocationFields(value, false); err != nil {
		return err
	}
	digest, err := invocationDigest(value)
	if err != nil {
		return err
	}
	if value.InvocationSHA256 != digest {
		return fmt.Errorf("invocation_sha256 does not match canonical preimage")
	}
	_, err = canonicalTyped(value, maxInvocationBytes)
	return err
}

// SealCapabilityInvocation seals one exact blank-identity invocation copy.
func SealCapabilityInvocation(value *CapabilityInvocation) (*CapabilityInvocation, error) {
	if value == nil || value.InvocationID != "" || value.InvocationSHA256 != "" {
		return nil, fmt.Errorf("sealing invocation requires blank identity fields")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := invocationDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.InvocationID, sealed.InvocationSHA256 = invocationPrefix+digest, digest
	return sealed, validateInvocation(sealed)
}

// DecodeCapabilityInvocation decodes one exact canonical sealed invocation.
func DecodeCapabilityInvocation(data []byte) (*CapabilityInvocation, error) {
	var value CapabilityInvocation
	if err := decodeTypedExact(data, maxInvocationBytes, &value); err != nil {
		return nil, err
	}
	return &value, validateInvocation(&value)
}

func eventDigest(value *InteractionEvent) (string, error) {
	blank, err := cloneValue(value)
	if err != nil {
		return "", err
	}
	blank.EventID, blank.EventSHA256 = "", ""
	if err := validateEventFields(blank, true); err != nil {
		return "", err
	}
	return typedDigest(blank, eventDomain, maxEventBytes)
}

func validateEvent(value *InteractionEvent) error {
	if err := validateEventFields(value, false); err != nil {
		return err
	}
	digest, err := eventDigest(value)
	if err != nil {
		return err
	}
	if value.EventSHA256 != digest {
		return fmt.Errorf("event_sha256 does not match canonical preimage")
	}
	_, err = canonicalTyped(value, maxEventBytes)
	return err
}

// SealInteractionEvent seals one exact blank-identity event copy.
func SealInteractionEvent(value *InteractionEvent) (*InteractionEvent, error) {
	if value == nil || value.EventID != "" || value.EventSHA256 != "" {
		return nil, fmt.Errorf("sealing event requires blank identity fields")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := eventDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.EventID, sealed.EventSHA256 = eventPrefix+digest, digest
	return sealed, validateEvent(sealed)
}

// DecodeInteractionEvent decodes one exact canonical sealed event.
func DecodeInteractionEvent(data []byte) (*InteractionEvent, error) {
	var value InteractionEvent
	if err := decodeTypedExact(data, maxEventBytes, &value); err != nil {
		return nil, err
	}
	return &value, validateEvent(&value)
}

func executionReceiptDigest(value *ExecutionReceipt) (string, error) {
	blank, err := cloneValue(value)
	if err != nil {
		return "", err
	}
	blank.ExecutionReceiptID, blank.ExecutionReceiptSHA256 = "", ""
	if err := validateExecutionReceiptFields(blank, true); err != nil {
		return "", err
	}
	return typedDigest(blank, executionReceiptDomain, maxExecutionReceiptBytes)
}

func validateExecutionReceipt(value *ExecutionReceipt) error {
	if err := validateExecutionReceiptFields(value, false); err != nil {
		return err
	}
	digest, err := executionReceiptDigest(value)
	if err != nil {
		return err
	}
	if value.ExecutionReceiptSHA256 != digest {
		return fmt.Errorf("execution_receipt_sha256 does not match canonical preimage")
	}
	_, err = canonicalTyped(value, maxExecutionReceiptBytes)
	return err
}

// SealExecutionReceipt seals one exact blank-identity execution receipt copy.
func SealExecutionReceipt(value *ExecutionReceipt) (*ExecutionReceipt, error) {
	if value == nil || value.ExecutionReceiptID != "" || value.ExecutionReceiptSHA256 != "" {
		return nil, fmt.Errorf("sealing ExecutionReceipt requires blank identity fields")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := executionReceiptDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.ExecutionReceiptID = executionReceiptPrefix + digest
	sealed.ExecutionReceiptSHA256 = digest
	return sealed, validateExecutionReceipt(sealed)
}

// DecodeExecutionReceipt decodes one exact canonical sealed execution receipt.
func DecodeExecutionReceipt(data []byte) (*ExecutionReceipt, error) {
	var value ExecutionReceipt
	if err := decodeTypedExact(data, maxExecutionReceiptBytes, &value); err != nil {
		return nil, err
	}
	return &value, validateExecutionReceipt(&value)
}

// DecodeArtifactRef decodes the exact reused ArtifactRef value object.
func DecodeArtifactRef(data []byte) (*ArtifactRef, error) {
	var value ArtifactRef
	if err := decodeTypedExact(data, maxArtifactRefBytes, &value); err != nil {
		return nil, err
	}
	return &value, validateArtifact(value, "ArtifactRef")
}

File: forge-runtime/crates/domain/src/kernel_operational_contract/records.rs
use sha2::{Digest, Sha256};

use super::{
    ArtifactReceipt, CapabilityInvocation, ExecutionReceipt, InteractionEvent,
    KernelOperationalContractError, MAX_ARTIFACT_RECEIPT_BYTES, MAX_EVENT_BYTES,
    MAX_EXECUTION_RECEIPT_BYTES, MAX_INVOCATION_BYTES,
    constants::{
        ARTIFACT_RECEIPT_DOMAIN, ARTIFACT_RECEIPT_PREFIX, EVENT_DOMAIN, EVENT_PREFIX,
        EXECUTION_RECEIPT_DOMAIN, EXECUTION_RECEIPT_PREFIX, INVOCATION_DOMAIN, INVOCATION_PREFIX,
    },
    invalid,
    validation::{
        validate_artifact_receipt_body, validate_event_body, validate_execution_receipt_body,
        validate_invocation_body,
    },
    wire,
};

fn digest(domain: &[u8], canonical: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(domain);
    hasher.update(canonical.as_bytes());
    crate::governance_contract::codec::lower_hex(&hasher.finalize())
}

fn artifact_receipt_digest(
    value: &ArtifactReceipt,
) -> Result<String, KernelOperationalContractError> {
    let mut blank = value.clone();
    blank.artifact_receipt_id.clear();
    blank.artifact_receipt_sha256.clear();
    validate_artifact_receipt_body(&blank, true)?;
    let canonical = wire::canonical_typed_with_max(&blank, MAX_ARTIFACT_RECEIPT_BYTES)?;
    Ok(digest(ARTIFACT_RECEIPT_DOMAIN, &canonical))
}

fn validate_artifact_receipt(
    value: &ArtifactReceipt,
) -> Result<(), KernelOperationalContractError> {
    validate_artifact_receipt_body(value, false)?;
    if value.artifact_receipt_sha256 != artifact_receipt_digest(value)? {
        return Err(invalid(
            "artifact_receipt_sha256 does not match the canonical preimage",
        ));
    }
    wire::canonical_typed_with_max(value, MAX_ARTIFACT_RECEIPT_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity `ArtifactReceipt` copy.
///
/// # Errors
///
/// Returns an error for a nonblank identity or invalid field, bound, order, or seal.
pub fn seal_artifact_receipt(
    value: &ArtifactReceipt,
) -> Result<ArtifactReceipt, KernelOperationalContractError> {
    if !value.artifact_receipt_id.is_empty() || !value.artifact_receipt_sha256.is_empty() {
        return Err(invalid(
            "sealing ArtifactReceipt requires blank identity fields",
        ));
    }
    let digest = artifact_receipt_digest(value)?;
    let mut sealed = value.clone();
    sealed.artifact_receipt_id = format!("{ARTIFACT_RECEIPT_PREFIX}{digest}");
    sealed.artifact_receipt_sha256 = digest;
    validate_artifact_receipt(&sealed)?;
    Ok(sealed)
}

/// Decodes one exact canonical sealed `ArtifactReceipt`.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, or invalid input.
pub fn decode_artifact_receipt(
    bytes: &[u8],
) -> Result<ArtifactReceipt, KernelOperationalContractError> {
    let value = wire::decode_typed(bytes, MAX_ARTIFACT_RECEIPT_BYTES)?;
    validate_artifact_receipt(&value)?;
    Ok(value)
}

fn invocation_digest(
    value: &CapabilityInvocation,
) -> Result<String, KernelOperationalContractError> {
    let mut blank = value.clone();
    blank.invocation_id.clear();
    blank.invocation_sha256.clear();
    validate_invocation_body(&blank, true)?;
    let canonical = wire::canonical_typed_with_max(&blank, MAX_INVOCATION_BYTES)?;
    Ok(digest(INVOCATION_DOMAIN, &canonical))
}

fn validate_invocation(value: &CapabilityInvocation) -> Result<(), KernelOperationalContractError> {
    validate_invocation_body(value, false)?;
    if value.invocation_sha256 != invocation_digest(value)? {
        return Err(invalid(
            "invocation_sha256 does not match the canonical preimage",
        ));
    }
    wire::canonical_typed_with_max(value, MAX_INVOCATION_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity `CapabilityInvocation` copy.
///
/// # Errors
///
/// Returns an error for a nonblank identity or invalid field, bound, order, or seal.
pub fn seal_capability_invocation(
    value: &CapabilityInvocation,
) -> Result<CapabilityInvocation, KernelOperationalContractError> {
    if !value.invocation_id.is_empty() || !value.invocation_sha256.is_empty() {
        return Err(invalid("sealing invocation requires blank identity fields"));
    }
    let digest = invocation_digest(value)?;
    let mut sealed = value.clone();
    sealed.invocation_id = format!("{INVOCATION_PREFIX}{digest}");
    sealed.invocation_sha256 = digest;
    validate_invocation(&sealed)?;
    Ok(sealed)
}

/// Decodes one exact canonical sealed `CapabilityInvocation`.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, or invalid input.
pub fn decode_capability_invocation(
    bytes: &[u8],
) -> Result<CapabilityInvocation, KernelOperationalContractError> {
    let value = wire::decode_typed(bytes, MAX_INVOCATION_BYTES)?;
    validate_invocation(&value)?;
    Ok(value)
}

fn event_digest(value: &InteractionEvent) -> Result<String, KernelOperationalContractError> {
    let mut blank = value.clone();
    blank.event_id.clear();
    blank.event_sha256.clear();
    validate_event_body(&blank, true)?;
    let canonical = wire::canonical_typed_with_max(&blank, MAX_EVENT_BYTES)?;
    Ok(digest(EVENT_DOMAIN, &canonical))
}

fn validate_event(value: &InteractionEvent) -> Result<(), KernelOperationalContractError> {
    validate_event_body(value, false)?;
    if value.event_sha256 != event_digest(value)? {
        return Err(invalid(
            "event_sha256 does not match the canonical preimage",
        ));
    }
    wire::canonical_typed_with_max(value, MAX_EVENT_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity `InteractionEvent` copy.
///
/// # Errors
///
/// Returns an error for a nonblank identity or invalid field, bound, order, or seal.
pub fn seal_interaction_event(
    value: &InteractionEvent,
) -> Result<InteractionEvent, KernelOperationalContractError> {
    if !value.event_id.is_empty() || !value.event_sha256.is_empty() {
        return Err(invalid("sealing event requires blank identity fields"));
    }
    let digest = event_digest(value)?;
    let mut sealed = value.clone();
    sealed.event_id = format!("{EVENT_PREFIX}{digest}");
    sealed.event_sha256 = digest;
    validate_event(&sealed)?;
    Ok(sealed)
}

/// Decodes one exact canonical sealed `InteractionEvent`.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, or invalid input.
pub fn decode_interaction_event(
    bytes: &[u8],
) -> Result<InteractionEvent, KernelOperationalContractError> {
    let value = wire::decode_typed(bytes, MAX_EVENT_BYTES)?;
    validate_event(&value)?;
    Ok(value)
}

fn execution_receipt_digest(
    value: &ExecutionReceipt,
) -> Result<String, KernelOperationalContractError> {
    let mut blank = value.clone();
    blank.execution_receipt_id.clear();
    blank.execution_receipt_sha256.clear();
    validate_execution_receipt_body(&blank, true)?;
    let canonical = wire::canonical_typed_with_max(&blank, MAX_EXECUTION_RECEIPT_BYTES)?;
    Ok(digest(EXECUTION_RECEIPT_DOMAIN, &canonical))
}

fn validate_execution_receipt(
    value: &ExecutionReceipt,
) -> Result<(), KernelOperationalContractError> {
    validate_execution_receipt_body(value, false)?;
    if value.execution_receipt_sha256 != execution_receipt_digest(value)? {
        return Err(invalid(
            "execution_receipt_sha256 does not match the canonical preimage",
        ));
    }
    wire::canonical_typed_with_max(value, MAX_EXECUTION_RECEIPT_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity `ExecutionReceipt` copy.
///
/// # Errors
///
/// Returns an error for a nonblank identity or invalid field, bound, order, or seal.
pub fn seal_execution_receipt(
    value: &ExecutionReceipt,
) -> Result<ExecutionReceipt, KernelOperationalContractError> {
    if !value.execution_receipt_id.is_empty() || !value.execution_receipt_sha256.is_empty() {
        return Err(invalid(
            "sealing ExecutionReceipt requires blank identity fields",
        ));
    }
    let digest = execution_receipt_digest(value)?;
    let mut sealed = value.clone();
    sealed.execution_receipt_id = format!("{EXECUTION_RECEIPT_PREFIX}{digest}");
    sealed.execution_receipt_sha256 = digest;
    validate_execution_receipt(&sealed)?;
    Ok(sealed)
}

/// Decodes one exact canonical sealed `ExecutionReceipt`.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, or invalid input.
pub fn decode_execution_receipt(
    bytes: &[u8],
) -> Result<ExecutionReceipt, KernelOperationalContractError> {
    let value = wire::decode_typed(bytes, MAX_EXECUTION_RECEIPT_BYTES)?;
    validate_execution_receipt(&value)?;
    Ok(value)
}

pub(super) fn validate_records(
    artifact_receipts: &[ArtifactReceipt],
    invocations: &[CapabilityInvocation],
    events: &[InteractionEvent],
    receipts: &[ExecutionReceipt],
) -> Result<(), KernelOperationalContractError> {
    for value in artifact_receipts {
        validate_artifact_receipt(value)?;
    }
    for value in invocations {
        validate_invocation(value)?;
    }
    for value in events {
        validate_event(value)?;
    }
    for value in receipts {
        validate_execution_receipt(value)?;
    }
    Ok(())
}
